src/ripple/app/misc/NetworkOPs.cpp

```diff
@@ -3117,6 +3286,296 @@ NetworkOPsImp::unsubAccountInternal(
+                {
+                    JLOG(m_journal.trace()) << "AccountHistory job for account "
+                                            << toBase58(accountId)
+                                            << " done, reached genesis ledger.";
+                    return;
+                }
+            }
+        });
+}
+
+void
+NetworkOPsImp::subAccountHistoryStart(
+    std::shared_ptr<ReadView const> const& ledger,
+    SubAccountHistoryInfoWeak& subInfo)
+{
+    subInfo.index_->separationLedgerSeq_ = ledger->seq();
+    auto const& accountId = subInfo.index_->accountId_;
+    auto const accountKeylet = keylet::account(accountId);
+    if (!ledger->exists(accountKeylet))
+    {
+        JLOG(m_journal.debug())
+            << "subAccountHistoryStart, no account " << toBase58(accountId)
+            << ", no need to add AccountHistory job.";
+        return;
+    }
+    if (accountId == genesisAccountId)
+    {
+        if (auto const sleAcct = ledger->read(accountKeylet); sleAcct)
+        {
+            if (sleAcct->getFieldU32(sfSequence) == 1)
+            {
+                JLOG(m_journal.debug())
+                    << "subAccountHistoryStart, genesis account "
+                    << toBase58(accountId)
+                    << " does not have tx, no need to add AccountHistory job.";
+                return;
+            }
+        }
+        else
+        {
+            assert(false);
+            return;
+        }
+    }
+    subInfo.index_->historyLastLedgerSeq_ = ledger->seq();
+    subInfo.index_->haveHistorical_ = true;
+
+    JLOG(m_journal.debug())
+        << "subAccountHistoryStart, add AccountHistory job: accountId="
+        << toBase58(accountId) << ", currentLedgerSeq=" << ledger->seq();
+
+    addAccountHistoryJob(subInfo);
+}
+
+error_code_i
+NetworkOPsImp::subAccountHistory(
+    InfoSub::ref isrListener,
+    AccountID const& accountId)
+{
+    if (!isrListener->insertSubAccountHistory(accountId))
+    {
+        JLOG(m_journal.debug())
+            << "subAccountHistory, already subscribed to account "
+            << toBase58(accountId);
+        return rpcINVALID_PARAMS;
+    }
+
+    std::lock_guard sl(mSubLock);
+    SubAccountHistoryInfoWeak ahi{
+        isrListener, std::make_shared<SubAccountHistoryIndex>(accountId)};
+    auto simIterator = mSubAccountHistory.find(accountId);
+    if (simIterator == mSubAccountHistory.end())
+    {
+        hash_map<std::uint64_t, SubAccountHistoryInfoWeak> inner;
+        inner.emplace(isrListener->getSeq(), ahi);
+        mSubAccountHistory.insert(
+            simIterator, std::make_pair(accountId, inner));
+    }
+    else
+    {
+        simIterator->second.emplace(isrListener->getSeq(), ahi);
+    }
+
+    auto const ledger = app_.getLedgerMaster().getValidatedLedger();
+    if (ledger)
+        subAccountHistoryStart(ledger, ahi);
+    else
+        JLOG(m_journal.debug())
+            << "subAccountHistory, no validated ledger yet, delay start";
```

Comment:
What actually happens in this case? Is the subscription successful? Maybe add a comment explaining what happens in this case.

Reply:
Thanks, comment added.